libtenzir/src/curl.cpp

```diff
@@ -0,0 +1,173 @@
+  TENZIR_ASSERT(not err);
+  // Enable all supported built-in compressions by setting the empty string.
+  // This can always be overriden by manually setting the Accept-Encoding
+  // header.
+  err = set(CURLOPT_ACCEPT_ENCODING, "");
+  TENZIR_ASSERT(not err);
+  err = to_error(curl_multi_add_handle(multi_, easy_));
+  TENZIR_ASSERT(not err);
+}
+
+curl::~curl() {
+  if (headers_)
+    curl_slist_free_all(headers_);
+  curl_multi_remove_handle(multi_, easy_);
+  curl_easy_cleanup(easy_);
+  curl_multi_cleanup(multi_);
+}
+
+auto curl::set(const curl_options& opts) -> caf::error {
+  if (not opts.default_protocol.empty())
+    if (auto err = set(CURLOPT_DEFAULT_PROTOCOL, opts.default_protocol.c_str()))
+      return err;
+  if (not opts.url.empty())
+    if (auto err = set(CURLOPT_URL, opts.url.c_str()))
+      return err;
+  if (opts.default_protocol == "http" || opts.default_protocol == "https") {
+    if (not opts.http.body.empty() || opts.http.method == "POST") {
+      // Setting the size manually is important because curl would otherwise
+      // use strlen to determine the body size. We may pass binary data at
+      // some point.
+      auto size = detail::narrow_cast<long>(opts.http.body.size());
+      if (auto err = set(CURLOPT_POSTFIELDSIZE, size))
+        return err;
+      // Here is choice of CURLOPT_POSTFIELDS vs. CURLOPT_COPYPOSTFIELDS,
+      // where the latter copies the input. We're copying now, but should
+      // revisit this for large POST bodies.
+      if (auto err = set(CURLOPT_COPYPOSTFIELDS, opts.http.body.c_str()))
+        return err;
+      // NB: setting body data via CURLOPT_POSTFIELDS implicitly sets the
+      // Content-Type header to 'application/x-www-form-urlencoded' unless we
+      // provide a Content-Type header that overrides it.
+    }
+    if (opts.http.method == "GET") {
+      if (auto err = set(CURLOPT_HTTPGET, 1L))
+        return err;
+    } else if (opts.http.method == "POST") {
+      if (auto err = set(CURLOPT_POST, 1L))
+        return err;
+    } else if (opts.http.method == "HEAD") {
+      if (auto err = set(CURLOPT_NOBODY, 1L))
+        return err;
+    } else if (not opts.http.method.empty()) {
+      if (auto err = set(CURLOPT_CUSTOMREQUEST, opts.http.method.c_str()))
+        return err;
+    }
+    if (not opts.http.headers.empty()) {
+      if (headers_) {
+        curl_slist_free_all(headers_);
+        headers_ = nullptr;
+      }
+      for (const auto& [header, value] : opts.http.headers) {
+        auto str = header;
+        str += ':';
+        if (not value.empty()) {
+          str += ' ';
+          str += value;
+        }
+        headers_ = curl_slist_append(headers_, str.c_str());
+      }
+    }
+    if (auto err = set(CURLOPT_HTTPHEADER, headers_))
+      return err;
+  }
+  return {};
+}
+
+auto curl::download(std::chrono::milliseconds timeout) -> generator<chunk_ptr> {
+  // curl_easy_setopt(easy_, CURLOPT_VERBOSE, 1L);
```

Comment:
I'm almost inclined to add an undocumented `--debug` flag. The output on stderr is incredibly valuable.

Any objections?

Reply:
Fine for now - the endgame for me would be having this verbose (and maybe also `s3` and `gcs` log) output as part of `TENZIR_VERBOSE`.

Reply:
Okay. It's [possible](https://curl.se/libcurl/c/CURLOPT_STDERR.html) to redirect to a `FILE*` instead of stderr, but that would require some interfacing to our logging framework.

I'll go with `--debug` for now.

Reply:
2d0e376a1b